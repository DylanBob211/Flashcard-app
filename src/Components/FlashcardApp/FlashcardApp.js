import React, { useState, useEffect } from 'react'
import { addNewList as addNewListDependent, deleteList as deleteListDependent} from '../../Actions/listActions' 
import { addWord as addWordDependent , deleteWord as deleteWordDependent } from '../../Actions/wordActions'
import Lists from '../Lists'
import ListForm from '../ListForm'


const FlashcardApp = () => {
    
    
    const [lists, setLists] = useState([
        {name : 'Rifiuti', words: [{ word :'secchio della spazzatura', url : 'No Img Available' }, { word :'discarica', url : 'No Img Available' }]},
        {name : 'Cibo', words: [{ word :'pizza', url : 'No Img Available' }, { word :'insalata', url : 'No Img Available' }]},
        {name : 'Viaggiare', words: [{ word :'maremma', url : 'No Img Available' }, { word :'colcane', url : 'No Img Available' }]}
    ])
    
    useEffect(()=> {
        console.log(lists)
    }, [lists])

    /*word Actions with dependencies injected*/
    
    const addWord = (newWord, id) => addWordDependent(newWord, setLists, id);
 
    const deleteWord = (wordName, id) => deleteWordDependent(wordName, setLists, id)

    /**********************/

    /*List Actions with dependencies injected*/

    const addNewList = newList => addNewListDependent(newList, setLists)
    const deleteList = listName => deleteListDependent(listName, setLists)

    /*******************/
    
    return (
        <div>
            <ListForm addNewList={addNewList} />
            <Lists 
                addWord={ addWord } 
                deleteWord={ deleteWord } 
                deleteList={ deleteList } 
                lists={ lists } 
            />
        </div>
    )
}

export default FlashcardApp
