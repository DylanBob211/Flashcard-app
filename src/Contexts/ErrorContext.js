import React, { useState, useContext } from 'react';

const ErrorContext = React.createContext(['', () => {}]);

export const useErrorContext = () => useContext(ErrorContext);

const ErrorProvider = ({ children }) => {
  const [error, setError] = useState('');
  return (
    <ErrorContext.Provider
      value={[
        error,
        setError,
      ]}
    >
      {children}
    </ErrorContext.Provider>
  );
};

export default ErrorProvider;
