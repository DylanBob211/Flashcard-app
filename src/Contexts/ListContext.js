import React, { useContext } from 'react';
import initialState from '../Components/initialState';
import useLists from '../Hooks/useLists';

const ListContext = React.createContext([initialState, () => {}]);

export const useListContext = () => useContext(ListContext);

const ListProvider = ({ children }) => {
  const { lists, setLists, addNewList } = useLists();
  return (
    <ListContext.Provider value={{ lists, setLists, addNewList }}>
      {children}
    </ListContext.Provider>
  );
};

export default ListProvider;
