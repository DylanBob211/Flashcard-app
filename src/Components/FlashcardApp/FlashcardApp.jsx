import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useErrorContext } from '../../Contexts/ErrorContext';
import { useListContext } from '../../Contexts/ListContext';
import ListCollection from '../ListCollection/ListCollection';
import PracticeWindow from '../PracticeWindow/PracticeWindow';
import ErrorModal from '../ErrorModal/ErrorModal';

const FlashcardApp = ({ languages }) => {
  const [windowState, setWindowState] = useState({ case: '', data: [] });
  const [error] = useErrorContext();
  const { lists } = useListContext();
  
  /* handlers */
  const getFlashcardData = (listId, wordItem) => lists.filter(list => list.id === listId)[0]
    .words.filter(word => word.word === wordItem.word)[0];

  const openFlashcard = listId => (wordItem) => {
    const flashcardData = getFlashcardData(listId, wordItem);
    setWindowState({ case: 'flashcard', data: flashcardData });
  };

  const getListData = listId => lists.filter(list => list.id === listId)[0];

  const openExerciseWindow = listId => () => {
    const listToPracticeData = getListData(listId);
    setWindowState({ case: 'list', data: listToPracticeData });
  };
  const closeExerciseWindow = () => {
    setWindowState(state => ({ ...state, case: '' }));
  };

  return (
    <div data-test="FlashcardAppContainer">
      <ErrorModal text={error} />
      <ListCollection
        lists={lists}
        getFlashcardData={getFlashcardData}
        openFlashcard={openFlashcard}
        openExerciseWindow={openExerciseWindow}
      />
      <PracticeWindow
        windowState={windowState}
        closeExerciseWindow={closeExerciseWindow}
      />
    </div>
  );
};

export default FlashcardApp;

FlashcardApp.propTypes = {
  languages: PropTypes.shape({
    from: PropTypes.string,
    to: PropTypes.string,
  }),
};

FlashcardApp.defaultProps = {
  languages: {
    from: 'en',
    to: 'it',
  },
};
