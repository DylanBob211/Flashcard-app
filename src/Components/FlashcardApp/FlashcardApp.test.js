import React from 'react';
import { mount, shallow } from 'enzyme';
import * as ErrorContext from '../../Contexts/ErrorContext';
import FlashcardApp from './FlashcardApp';

describe('FlashcardApp', () => {
  let mountedFlashcardAppComponent;
  const props = {
    languages: { from: '', to: '' },
  };

  const useErrorMock = jest.spyOn(ErrorContext, 'useErrorContext');
  const setErrorMock = jest.fn();
  let errorMock;
  useErrorMock.mockImplementation(() => [errorMock, setErrorMock]);

  const flashcardApp = ({ languages }, isShallow = false) => {
    if (isShallow) {
      return shallow(
        <FlashcardApp
          languages={languages}
        />,
      );
    }
    return mount(
      <FlashcardApp
        languages={languages}
      />,
    );
  };

  beforeEach(() => {
    mountedFlashcardAppComponent = undefined;
    jest.clearAllMocks();
  });

  beforeEach(() => {
    mountedFlashcardAppComponent = flashcardApp(props);
    errorMock = 'hello error';
  });

  it('matches the snapshot', () => {
    const snapshotComponent = flashcardApp(props, true);
    expect(snapshotComponent).toMatchSnapshot();
  });

  it('always render a div container', () => {
    const divContainer = mountedFlashcardAppComponent
      .find('[data-test="FlashcardAppContainer"]');
    expect(divContainer.exists()).toEqual(true);
  });

  it('always renders an ErrorModal component with a message if error is passed from context', () => {
    const errorModal = mountedFlashcardAppComponent
      .find('ErrorModal');
    expect(errorModal.exists()).toEqual(true);
  });

  it('passes the error text from context to the ErrorModal', () => {
    const errorModal = mountedFlashcardAppComponent
      .find('ErrorModal');
    expect(errorModal.props().text).toEqual('hello error');
  });

  it('renders a PracticeWindow component', () => {
    const practiceWindow = mountedFlashcardAppComponent
      .find('PracticeWindow');
    expect(practiceWindow.exists()).toEqual(true);
  });
});
