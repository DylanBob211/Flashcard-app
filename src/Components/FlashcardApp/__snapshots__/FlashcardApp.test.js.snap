// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashcardApp matches the snapshot 1`] = `
<div
  data-test="FlashcardAppContainer"
>
  <ErrorModal
    text="hello error"
  />
  <ListCollection
    getFlashcardData={[Function]}
    lists={Array []}
    openExerciseWindow={[Function]}
    openFlashcard={[Function]}
  />
  <PracticeWindow
    closeExerciseWindow={[Function]}
    windowState={
      Object {
        "case": "",
        "data": Array [],
      }
    }
  />
</div>
`;
